ethereumjs/ethereumjs-vm
# Blockchain: internal refactor

Closes #879 

This PR is a bit all over the place, here is what it currently does/fixes:

- Create an explicit test for blockchain reorgs, 221c07792e686dec6fac8039d016fc14cc186324
- Re-check the skipped tests in VM (these were "skipped" due to what it seemed to be a lack of reorg support, which actually was supported in the first place (😅 )). Hurrah! Lots of skipped tests now work! ca15b65b6be2d55e7e4f39ea0a3b0155467d7577
- Create a static method on `block`, `getCanonicalDifficulty`. Otherwise, one needs to create a mock header to call canonicalDifficulty on. 4967076b7f4a72e189fc18d1a6be3a7ce356cd2f
- Mark functions which were intended as private, `private`
- Fix race condition on Blockchain getters and the chain iterator (one can iterate the chain, then reorg, but this would mess up the chain iterator)
- Write a bit more comments and documentation of most functions
- If a non-canonical Genesis block is used (w.r.t. provided `Common`) then this should be provided in the `constructor` of `Blockchain`: if we would not do this, it is possible to overwrite the Genesis block (!)

Refactor: DB operations like calculating the keys of the DB are now handled as a `DatabaseOperation` object, so we don't have all these keys over the blockchain file. 

TODO: 

- Figure out why we have to `release` twice in `initAndLock`
- Write `initAndLock` tests (concurrency)
- Check examples in VM w.r.t. `genesis` blocks
- Make the introduced `static` methods in `Block` non-static.
